### DISCUSSION TITLE
mesh= Gmsh3D("Model3D.geo") can not get the mesh

### DISCUSSION THREAD

**Discussion by LiuDongDong1989** [usnistgov/fipy#964] [Q&A] [answered]
Model3D.geo
```
SetFactory("OpenCASCADE");

Mesh.CharacteristicLengthMax = 0.1;
Mesh.CharacteristicLengthMin = 0.1;

Box(1) = {0, 0, 0, 0.5, 0.3, 0.2};

angle = 22; 
Rotate {{0,1,0}, {0,0,0}, angle*Pi/180} {
    Volume{1}; 
}

Physical Surface("TOP") = {6};
```
In Gmsh, the above code can obtain the 3D mesh. 
However, In fipy, `mesh= Gmsh3D("Model3D.geo")` can not work. If I DELETE `Physical Surface("TOP") = {6};`, this code can work in FIPY.
The error:  
 ```
 raise GmshException(errStr)
fipy.meshes.gmshMesh.GmshException: Gmsh hasn't produced any cells! Check your Gmsh code.
```

Besides, if I used the "Model3D.msh" always can not work.

My version= fipy 3.4.4 + gmsh 4.11.1 +MacOS

**Comment by guyer** [comment, marked as answer]
If you compare the `.msh` file generated with and without `Physical Surface("TOP") = {6};`, you can see that when this line is present, the resulting MSH file consists only of type-2 elements, which are triangles. There are no 3D elements (tetrahedra) in the file and so FiPy is unable to create any cells. This isn't a FiPy issue; it's a Gmsh issue.

The Gmsh documentation isn't very clear about this, although it's sort of explained in [FAQ C.5.3](https://gmsh.info/doc/texinfo/gmsh.html#Mesh-module-questions):
> By default, if physical groups are defined, the output mesh only contains those elements that belong to physical entities. So to save only 3D elements, simply define one (or more) physical volume(s) and don’t define any physical surfaces, physical curves or physical points.

So, if you add, e.g.,
```
Physical Volume("everythang") = {1};
```
to your `.geo` file, then Gmsh will include 3D elements in the output MSH file and FiPy will be able to read it.

There is a discussion about this in the docstring for [`Gmsh2D`](https://pages.nist.gov/fipy/en/modernize_sphinx/generated/fipy.meshes.gmshMesh.html#fipy.meshes.gmshMesh.Gmsh2D), but it could be clearer, and it needs to be repeated for `Gmsh3D`.
